python/lsst/ip/diffim/imageDecorrelation.py

```diff
@@ -0,0 +1,191 @@
+    @return sigma-clipped mean and std of input array
+    """
+    _, low, upp = sigmaclip(im, low=low, high=high)
+    tmp = im[(im > low) & (im < upp)]
+    mean1 = np.nanmean(tmp)
+    sig1 = np.nanstd(tmp)
+    return mean1, sig1
+
+
+def getImageGrid(im):
+    """! Convenience function for creating a grid of pixel coordinates covering a 2-d array,
+    using `numpy.meshgrid()`
+    @param im input 2-d numpy array
+    @return two numpy arraysm each with same shape as `im`, containing x- and y- coordinate
+    grid.
+    """
+    xim = np.arange(np.int(-np.floor(im.shape[0]/2.)), np.int(np.floor(im.shape[0]/2)))
+    yim = np.arange(np.int(-np.floor(im.shape[1]/2.)), np.int(np.floor(im.shape[1]/2)))
+    x0im, y0im = np.meshgrid(xim, yim)
+    return x0im, y0im
+
+
+def computeDecorrelationCorrectionKernel(kappa, sig1=0.2, sig2=0.2):
+    """! Compute the Lupton/ZOGY post-conv. kernel for decorrelating an
+    image difference, based on the PSF-matching kernel.
+    @param kappa  A matching kernel array derived from Alard & Lupton PSF matching
+    @param sig1   Average sqrt(variance) of template image used for PSF matching
+    @param sig2   Average sqrt(variance) of science image used for PSF matching
+    @return a 2-d numpy.array containing the correction kernel
+
+    @note As currently implemented, kappa is a static (single) kernel.
+    """
+    def post_conv_kernel_ft2(kernel, sig1=1., sig2=1.):
+        kft = fft2(kernel)
+        return np.sqrt((sig1**2 + sig2**2) / (sig1**2 + sig2**2 * kft**2))
+
+    def post_conv_kernel2(kernel, sig1=1., sig2=1.):
+        kft = post_conv_kernel_ft2(kernel, sig1, sig2)
+        out = ifft2(kft)
+        return out
+
+    pck = post_conv_kernel2(kappa, sig1=sig2, sig2=sig1)
+    pck = ifftshift(pck.real)
+
+    # I think we may need to "reverse" the PSF, as in the ZOGY (and Kaiser) papers...
+    # This is the same as taking the complex conjugate in Fourier space before FFT-ing back to real space.
+    if False:  # TBD: figure this out. For now, we are turning it off.
+        pck = pck[::-1, :]
+
+    return pck
+
+
+def computeCorrectedDiffimPsf(kappa, im2_psf, sig1=0.2, sig2=0.2):
+    """! Compute the (decorrelated) difference image's new PSF:
+    post_conv_psf = phi_1(k) * sym.sqrt((sig1**2 + sig2**2) / (sig1**2 + sig2**2 * kappa_ft(k)**2))
+
+    @param kappa  A matching kernel array derived from Alard & Lupton PSF matching
+    @param im2_psf The uncorrected psf array of the science image (and also of the diffim)
+    @param sig1   Average sqrt(variance) of template image used for PSF matching
+    @param sig2   Average sqrt(variance) of science image used for PSF matching
+    @return a 2-d numpy.array containing the new PSF
+    """
+    def post_conv_psf_ft2(psf, kernel, sig1=1., sig2=1.):
+        # Pad psf or kernel symmetrically to make them the same size!
+        if psf.shape[0] < kernel.shape[0]:
+            while psf.shape[0] < kernel.shape[0]:
```

Comment:
Why use a loop instead of simply doing all the padding at once? If that is truly the best way then an explanatory comment would be appreciated.